File: 0150_evaluate_reverse_polish_notation/solution01.cpp
#include <cmath>
#include <cstdlib>
#include <stack>
#include <unordered_map>
#include <unordered_set>

int add_(int x, int y) { return x + y; }
int sub_(int x, int y) { return x - y; }
int mul_(int x, int y) { return x * y; }
int div_(int x, int y) { return trunc(x / y); }

// runtime: 37.73%, memory: 15.04% 
class Solution {
public:
    int evalRPN(vector<string>& tokens) {
        unordered_set<string> operators = {"+", "-", "*", "/"};
        unordered_map<string, int(*)(int, int)> operations = {
            { "+", add_ }, 
            { "-", sub_ }, 
            { "*", mul_ }, 
            { "/", div_ },
        };
        stack<int> st;

        for (const string& tok : tokens) {
            // if (operators.find(tok) != operators.end()) {
            if (operators.contains(tok)) {
                int y = st.top(); st.pop();
                int x = st.top(); st.pop();

                int result = operations[tok](x, y);
                st.push(result);
            } else {
                st.push(stoi(tok));
            }
        }

        return st.top();
    }
};

File: 0150_evaluate_reverse_polish_notation/solution01.py
import math 
from typing import List 

# 28 Jan 2025
# runtime: 57.15%, memory: 100.00% 
class Solution:
    def evalRPN(self, tokens: List[str]) -> int:
        stack: List[str] = []
        oper_funcs = {
            "+": lambda x, y: x + y,
            "-": lambda x, y: x - y,
            "*": lambda x, y: x * y,
            "/": lambda x, y: math.trunc(x / y)
        }
        
        for token in tokens:
            if token in oper_funcs.keys():
                op2 = int(stack.pop())
                op1 = int(stack.pop())

                token = oper_funcs[token](op1, op2)
                
            stack.append(f"{token}")
            
        return int(stack.pop())
        
        